channels/offline-channel-js/src/OfflineChannel.ts

```diff
@@ -0,0 +1,604 @@
+                } else {
+                    _sendNextBatchTimer.refresh()
+                }
+            }
+
+            function _clearScheduledTimer() {
+                _inMemoFlushTimer && _inMemoFlushTimer.cancel();
+                _sendNextBatchTimer && _sendNextBatchTimer.cancel();
+                _inMemoFlushTimer = null;
+                _sendNextBatchTimer = null;
+            }
+
+            function _queueStorageEvent<T>(taskName: string, task: T | IPromise<T>) {
+                if (_taskScheduler) {
+                    _taskScheduler.queue(() => {
+                        return task;
+                    }, taskName).catch((reason) => {
+                        // Just handling any rejection to avoid an unhandled rejection event
+                    });
+                }
+                
+            }
+
+            function _setRetryTime(linearFactor: number = 1) {
+                const SlotDelayInSeconds = 10;
+                let delayInSeconds: number;
+        
+                if (_consecutiveErrors <= 1) {
+                    delayInSeconds = SlotDelayInSeconds;
+                } else {
+                    const backOffSlot = (Math.pow(2, _consecutiveErrors) - 1) / 2;
+                    // tslint:disable-next-line:insecure-random
+                    let backOffDelay = Math.floor(Math.random() * backOffSlot * SlotDelayInSeconds) + 1;
+                    backOffDelay = linearFactor * backOffDelay;
+                    delayInSeconds = Math.max(Math.min(backOffDelay, 3600), SlotDelayInSeconds);
+                }
+        
+                // TODO: Log the backoff time like the C# version does.
+                let retryAfterTimeSpan = dateNow() + (delayInSeconds * 1000);
+        
+                // TODO: Log the retry at time like the C# version does.
+                _retryAt = retryAfterTimeSpan;
+            }
+
+            function _getCoreItemCtx(coreConfig: IConfiguration & IConfig, core: IAppInsightsCore, extensions: IPlugin[], pluginChain?: ITelemetryPluginChain) {
+                if (coreConfig) {
+                    // Make sure the extensionConfig exists
+                    coreConfig.extensionConfig = coreConfig.extensionConfig || {};
+                }
+
+                if (!pluginChain && core) {
+                    // Get the first plugin from the core
+                    pluginChain = core.getProcessTelContext().getNext();
+                }
+
+                let nextPlugin: IPlugin = null;
+                let rootNext = _self._getTelCtx().getNext();
+                if (rootNext) {
+                    nextPlugin = rootNext.getPlugin();
+                }
+
+                return createProcessTelemetryContext(pluginChain, coreConfig, core, nextPlugin);
+            }
+
+
+            function _getPayload(evt: IPostTransmissionTelemetryItem | ITelemetryItem) {
+                try {
+
+                    if (evt) {
+                        evt.iKey = evt.iKey || _urlCfg.iKey;
```

Comment:
If the iKey is not set by the time we get here then there is a potential issue -- but we also talked about the "overrideIkey"  (but only within the `processTelemetry` -- not here (I don't think)